// Type: Homework
// Level: 5
// Section: 3.6 Exception Handling
// Exercise: 1
// Description: Definition file contains class Line()
// In the array class we created previously, the bounds checking was very basic.
// There was no error generated, but setting an element was ignored or the first element was returned.
// Obviously you want to know if there was an out of bounds error.
// This is possible by exception handling. Change the Array class to throw exceptions:
//  •In the GetElement(), SetElement() and index operator throw -1 if the index was too small or too big.
//  •In the main program, create an Array object and access an element that does not exist.
//      Run the program. What happens?
//  •The exception must be caught, so place the code inside a try ... catch block that catches an int.
//  •In the catch handler, print an error message.

/*---------------------------------*/
#ifndef Line_HPP
#define Line_HPP

#include <iostream>
#include <Point.hpp>
#include <Shape.hpp>

using namespace std;

/*---------------------------------*/
class Line : public Shape
{
private:
    Point P1;  // point 1
    Point P2;  // point 2

public:
    // Constructors
    Line();  // Default constructor
    Line(Point &newP1, Point &newP2);  // Initialize with start and end points
    Line(const Line &line);  // Copy constructor

    // Destructor
    ~Line();

    // Accessing functions
    Point point1() const;  // The start point
    void point1(const Point &newP1);

    Point point2() const;  // The end point
    void point2(const Point &newP2);

    string ToString() const;  // Return string description of the line
    double Length() const;  // Calculate the length of the line
    void Draw() const;  // Overidden function from Shape() simulating drawing by printing some text

    // Overloading Operator
    Line &operator=(const Line &source); // Assignment operator.
    friend ostream &operator<<(ostream &os, const Line &l); // Send to ostream.
};

#endif // Line_HPP